### PR TITLE
Support restricted index patterns in Elasticsearch log handler

### LINKED ISSUE #16828: Cannot Set Index Pattern on Elasticsearch as a Log Handler
**Apache Airflow version**: 2.0.0

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): ```Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:28:09Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.8-aliyun.1", GitCommit:"94f1dc8", GitTreeState:"", BuildDate:"2021-01-10T02:57:47Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}```

**Environment**: -

- **Cloud provider or hardware configuration**: Alibaba Cloud
- **OS** (e.g. from /etc/os-release): Debian GNU/Linux 10 (buster)
- **Kernel** (e.g. `uname -a`): `Linux airflow-webserver-fb89b7f8b-fgzvv 3.10.0-1160.11.1.el7.x86_64 #1 SMP Fri Dec 18 16:34:56 UTC 2020 x86_64 GNU/Linux`
- **Install tools**: Helm (Custom)
- **Others**: None

**What happened**:
My Airflow use fluent-bit to catch the stdout logs from airflow containers and then send the logs messages to Elasticsearch in a remote machine and it works well, I can see the logs through Kibana. But the Airflow cannot display the logs, because an error:
```
ERROR - Exception on /get_logs_with_metadata [GET]
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/airflow/.local/lib/python3.8/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/airflow/.local/lib/python3.8/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)  
  File "/home/airflow/.local/lib/python3.8/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/airflow/.local/lib/python3.8/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/airflow/.local/lib/python3.8/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/www/auth.py", line 34, in decorated
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/www/decorators.py", line 60, in wrapper
    return f(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/session.py", line 65, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/www/views.py", line 1054, in get_logs_with_metadata
    logs, metadata = task_log_reader.read_log_chunks(ti, try_number, metadata)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/log/log_reader.py", line 58, in read_log_chunks
    logs, metadatas = self.log_handler.read(ti, try_number, metadata=metadata)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/utils/log/file_task_handler.py", line 217, in read
    log, metadata = self._read(task_instance, try_number_element, metadata)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/elasticsearch/log/es_task_handler.py", line 160, in _read
    logs = self.es_read(log_id, offset, metadata)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/providers/elasticsearch/log/es_task_handler.py", line 233, in es_read
    max_log_line = search.count()
  File "/home/airflow/.local/lib/python3.8/site-packages/elasticsearch_dsl/search.py", line 701, in count
    return es.count(index=self._index, body=d, **self._params)["count"]
  File "/home/airflow/.local/lib/python3.8/site-packages/elasticsearch/client/utils.py", line 84, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/elasticsearch/client/__init__.py", line 528, in count
    return self.transport.perform_request(
  File "/home/airflow/.local/lib/python3.8/site-packages/elasticsearch/transport.py", line 351, in perform_request
    status, headers_response, data = connection.perform_request(
  File "/home/airflow/.local/lib/python3.8/site-packages/elasticsearch/connection/http_urllib3.py", line 261, in perform_request
    self._raise_error(response.status, raw_data)
  File "/home/airflow/.local/lib/python3.8/site-packages/elasticsearch/connection/base.py", line 181, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(
elasticsearch.exceptions.AuthorizationException: AuthorizationException(403, 'security_exception', 'no permissions for [indices:data/read/search] and User [name=airflow, backend_roles=[], request
```
but when I debug and use this code, I can see the logs:
```
es = elasticsearch.Elasticsearch(['...'], **es_kwargs)
es.search(index="airflow-*", body=dsl)
```

and when I look into the source code of elasticsearch providers there are no definition of the index-pattern on that

https://github.com/apache/airflow/blob/88199eefccb4c805f8d6527bab5bf600b397c35e/airflow/providers/elasticsearch/log/es_task_handler.py#L216

so I assume the issue is insufficient permission to scan all the indices, therefore, how can I set the index-pattern so that Airflow only reads certain indices?
Thank you!

**What you expected to happen**: The Airflow configuration has option to add elasticsearch index pattern so that airflow only queries certain indices, not querying all indexes on the elasticsearch server

**How to reproduce it**: Click log button on task popup modal to see logs page

**Anything else we need to know**: Every time etc

### PR DESCRIPTION
**Pull request by kouk**
Sometimes Airflow doesn't have the ability to search across all indices
in an Elasticsearch server. This might be due to security settings in
the server. In these cases fetching the remote logs fails. To fix this
we create a index_patterns configuration setting that can be set to a
more restrictive pattern.

closes: #16828

### CODE REVIEW

**Comment by kouk** [comment]
@jedcunningham hi! I am running airflow with this patch at work and it is solving the issue for us. Appreciate it if you could take a look.

File: airflow/config_templates/config.yml [hunk 1]
```diff
@@ -2246,6 +2246,13 @@
       type: string
       example: ~
       default: "offset"
+    - name: index_patterns
+      description: |
+        Comma separated list of index patterns to use when searching for logs (default: `_all`).
+      version_added: 2.3.0
+      type: string
+      example: something-*
+      default: "_all"
```
Comment (uranusjr) [inline airflow/config_templates/config.yml:2326]: Is this a good default? Since this handles wildcard expansion, perhaps we could simply make the default `*` or an empty string instead.

Comment (kouk) [reply airflow/config_templates/config.yml:2326]: @uranusjr indeed we could use `*`. According to the documentation it is equivalent to `_all`. See
https://www.elastic.co/guide/en/elasticsearch/reference/8.2/search-search.html#search-search-api-path-params

Initially I had it set to `*` and then changed it to `_all` because it seemed more readable. That said, I don't mind changing it back to `*` if ASF members prefer it.

Comment (uranusjr) [reply airflow/config_templates/config.yml:2326]: If this is a special case in Elastic, I am OK to use `_all` as the default. The main motivation behind using `*` is you don’t need a special case in `_validate_search_targets` for it. But if we need to handle `_all` specially anyway, the difference is minimal.

Comment (kouk) [reply airflow/config_templates/config.yml:2326]: For background:
https://github.com/elastic/elasticsearch-py/blob/d66dfcb9c7e2bc592392ca00fde62889d181c0af/elasticsearch/_sync/client/__init__.py#L1166

Comment (kouk) [reply airflow/config_templates/config.yml:2326]: @uranusjr I'm not 100% sure but I think that it's a very common Elasticsearch idiom to use `_all` or the empty string. Perhaps we could change the default to an empty string?

Comment (kouk) [reply airflow/config_templates/config.yml:2326]: in fact I i'm realising now that `FakeElasticsearch` should probably handle the empty string case.

File: tests/providers/elasticsearch/log/elasticmock/fake_elasticsearch.py [hunk 2]
```diff
@@ -362,6 +363,18 @@ def match_must_phrase(document, matches, must):
...
+    # Check index(es) exists
+    def _validate_search_targets(self, targets):
+        matches = set()
+        for target in targets:
+            if target == '_all':
+                matches.update(self.__documents_dict.keys())
+            elif '*' in target:
+                matches.update(fnmatch.filter(self.__documents_dict.keys(), target))
+            elif target not in self.__documents_dict:
+                raise NotFoundError(404, f'IndexMissingException[[{target}] missing]')
```
Comment (uranusjr) [inline tests/providers/elasticsearch/log/elasticmock/fake_elasticsearch.py:375]: The `keys()` calls are not needed.

Given that a wildcard match is ignored if it does not match anything, we probably should do the same about normal names.

```suggestion
            if target == '_all':
                matches.update(self.__documents_dic)
            else:
                matches.update(fnmatch.filter(self.__documents_dict, target))
```

Comment (uranusjr) [inline tests/providers/elasticsearch/log/elasticmock/fake_elasticsearch.py]: The `keys()` call are not needed.

Since an exception is raised when `target` does not exist, maybe we should do the same if a wildcard pattern does not match anything?

```suggestion
            if target == '_all':
                matches.update(self.__documents_dict)
                continue
            target_matches = fnmatch.filter(self.__documents_dict, target)
            if not target_matches:
                raise NotFoundError(404, f'IndexMissingException[[{target}] missing]')
            matches.update(target_matches)
```

Comment (kouk) [reply tests/providers/elasticsearch/log/elasticmock/fake_elasticsearch.py]: @uranusjr actually Elasticsearch treats the case of a wildcard pattern not matching differently. See the `allow_no_indices` query parameter: https://www.elastic.co/guide/en/elasticsearch/reference/8.2/search-search.html#search-search-api-query-params
tl;dr if you specify an index name _without a pattern_ this results in 404. If you specify an index _with a pattern_ then by default you do not get a 404 even if there are no matching indices.

Comment (kouk) [reply tests/providers/elasticsearch/log/elasticmock/fake_elasticsearch.py]: For example:

    $ curl 'localhost:9200/whateverman-*/_search'
    {"took":0,"timed_out":false,"_shards":{"total":0,"successful":0,"skipped":0,"failed":0},"hits":{"total":{"value":0,"relation":"eq"},"max_score":0.0,"hits":[]}}

but

    $ curl 'localhost:9200/whateverman-foobar/_search'
    {"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [whateverman-foobar]","resource.type":"index_or_alias","resource.id":"whateverman-foobar","index_uuid":"_na_","index":"whateverman-foobar"}],"type":"index_not_found_exception","reason":"no such index [whateverman-foobar]","resource.type":"index_or_alias","resource.id":"whateverman-foobar","index_uuid":"_na_","index":"whateverman-foobar"},"status":404}

Comment (kouk) [reply tests/providers/elasticsearch/log/elasticmock/fake_elasticsearch.py]: However I haven't bothered to implement the case where`allow_no_indices=false` (would be necessary if the provider implementation was changed to override the default behavior):

    $ curl 'localhost:9200/whateverman-*/_search?allow_no_indices=false'
    {"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index [whateverman-*]","resource.type":"index_or_alias","resource.id":"whateverman-*","index_uuid":"_na_","index":"whateverman-*"}],"type":"index_not_found_exception","reason":"no such index [whateverman-*]","resource.type":"index_or_alias","resource.id":"whateverman-*","index_uuid":"_na_","index":"whateverman-*"},"status":404}

Comment (uranusjr) [reply tests/providers/elasticsearch/log/elasticmock/fake_elasticsearch.py]: OK, let’s just fix the unnecessary `keys()` calls then.

Comment (kouk) [reply tests/providers/elasticsearch/log/elasticmock/fake_elasticsearch.py]: Done!

**Review by uranusjr — APPROVED** [review]

File: airflow/config_templates/config.yml [hunk 3]
```diff
@@ -2246,6 +2246,13 @@
       type: string
       example: ~
       default: "offset"
+    - name: index_patterns
+      description: |
+        Comma separated list of index patterns to use when searching for logs (default: `_all`).
+      version_added: 2.3.0
```
Comment (eladkal) [inline airflow/config_templates/config.yml]:
```suggestion
      version_added: 2.3.3
```

**Review by jedcunningham — CHANGES_REQUESTED** [review]
This needs some changes to support backwards compatibility. We can't set `index_patterns` in `airflow_local_settings`, because we might have an older provider. We also should try and get the config value when on older core versions. Look at how `es_kwargs` is handled - that is a good example.

Otherwise, this looks good.

File: airflow/config_templates/config.yml [hunk 4]
```diff
@@ -2246,6 +2246,13 @@
       type: string
       example: ~
       default: "offset"
+    - name: index_patterns
+      description: |
+        Comma separated list of index patterns to use when searching for logs (default: `_all`).
+      version_added: 2.3.3
```
Comment (jedcunningham) [inline airflow/config_templates/config.yml]:
```suggestion
      version_added: 2.4.0
```

Oh, and this will end up being in core 2.4.0, as this isn't a bug (though I'm glad to see it supported 🍺).

**Comment by kouk** [comment]
Yeah, I was on vacation..will look at this today.

**Comment by potiuk** [comment]
@jedcunningham - I think that one waits for you :)

**Comment by potiuk** [comment]
You will need to rebase @kouk to account for recent changes in main.

**Comment by potiuk** [comment]
Can you also correct the version according to @jedcunningham comment @kouk ?

**Comment by kouk** [comment]
@jedcunningham @potiuk I bumped the version to 2.4.1. Also I pushed acommit that updates the tests to emulate the case where we have an older version of airflow core, where no index pattern is passed by default to the constructor. Let me know if you prefer the previous version.

**Comment by potiuk** [comment]
> @jedcunningham @potiuk I bumped the version to 2.4.1. Also I pushed acommit that updates the tests to emulate the case where we have an older version of airflow core, where no index pattern is passed by default to the constructor. Let me know if you prefer the previous version.

I think it will have to be 2.5.0 - > we are following semver and we only add bugfixes to "patchlevel" versions. This one will have to wait for 2.5 to be released.

File: airflow/providers/elasticsearch/log/es_task_handler.py [hunk 5]
```diff
@@ -114,6 +116,7 @@ def __init__(
         self.json_fields = [label.strip() for label in json_fields.split(",")]
         self.host_field = host_field
         self.offset_field = offset_field
+        self.index_patterns = index_patterns or es_configs.get('index_patterns', '_all')
```
Comment (jedcunningham) [inline airflow/providers/elasticsearch/log/es_task_handler.py]:
```suggestion
        self.index_patterns = index_patterns or conf.get('elasticsearch', 'index_patterns', fallback='_all')
```

Then you can drop 98 (`es_configs`). Also, can't we set this default on 86, letting us do a simple assignment here?

**Comment by jedcunningham** [comment]
Marking this for 2.5.0, but after the bulk of these changes land in a provider release, it'll work if you set the config.

**Comment by kouk** [comment]
@jedcunningham thanks for pointing that out. I wasn't familiar enough with the config code to realize that it's ConfigParser under the hood. I've updated both the version and the config handling code like you suggested.

**Comment by kouk** [comment]
@eladkal @jedcunningham anything else we need here?

**Comment by kouk** [comment]
@jedcunningham appreciate if you could check if anything else is needed here.

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Rebased to be sure. @jedcunningham - want to still take a look?

File: airflow/config_templates/config.yml [hunk 6]
```diff
@@ -2317,6 +2317,13 @@
       type: string
       example: ~
       default: "offset"
+    - name: index_patterns
+      description: |
+        Comma separated list of index patterns to use when searching for logs (default: `_all`).
+      version_added: 2.5.0
```
Comment (jedcunningham) [inline airflow/config_templates/config.yml]:
```suggestion
      version_added: 2.6.0
```

**Review by jedcunningham — APPROVED** [review]

**Comment by jedcunningham** [comment]
Thanks @kouk! Congrats on your first commit 🎉. (I apologize for the delay)

As I mentioned earlier, since almost all of the changes are in the provider, this feature will be available once it's released in the provider (even before 2.6 with the config option is released).

**Comment by kouk** [comment]
thank you @jedcunningham  and everyone else.
